src/main/java/pro/beam/api/http/BeamHttpClient.java

```diff
@@ -166,25 +174,55 @@ private HttpEntity makeEntity(Object... args) {
         return new Callable<T>() {
             @Override public T call() throws IOException, HttpBadResponseException {
                 BeamHttpClient self = BeamHttpClient.this;
-
-                HttpResponse partialResponse = self.http.execute(request, self.context);
-                HttpCompleteResponse completeResponse = new HttpCompleteResponseHandler().handleResponse(partialResponse);
-
-                if (completeResponse.status.getStatusCode() >= 300) {
-                    throw new HttpBadResponseException(completeResponse);
-                }
-
-                // Allow a null response to be given back, such that we return a ListenableFuture
-                // with null.
-                if (type != null) {
-                    return self.beam.gson.fromJson(completeResponse.body(), type);
-                } else {
-                    return null;
+                try {
+                    return handleRequest(request, type);
+                } catch(HttpBadResponseException e) {
+                    if (e.response.status.getStatusCode() == CSRF_CODE) {
+                        return handleRequest(request, type);
+                    }
                 }
+                return null;
             }
         };
     }
 
+    public <T> T handleRequest(HttpUriRequest request, Class<T> type) throws IOException, HttpBadResponseException {
+        HttpResponse partialResponse = this.makeRequest(request);
+        return this.handleResponse(partialResponse, type);
+    }
+
+    public HttpResponse makeRequest(HttpUriRequest request) throws IOException {
+        return this.http.execute(request, this.context);
+    }
+
+    public <T> T handleResponse(HttpResponse partialResponse, Class<T> type) throws IOException, HttpBadResponseException {
+        HttpCompleteResponse completeResponse = new HttpCompleteResponseHandler().handleResponse(partialResponse);
+
+        this.handleCSRF(partialResponse);
+
+        if (completeResponse.status.getStatusCode() >= 300) {
+            throw new HttpBadResponseException(completeResponse);
```

Comment:
It would probably be good to:
 - follow redirects when they direct to another endpoint (so we can rename things in the future, like tetris -> interactive)
 - handle redirects some other way? Or just return them?